compiler/rustc_borrowck/src/region_infer/opaque_types.rs

```diff
@@ -78,6 +82,24 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                 }
             }
         }
+
+        if !infcx.tcx.is_typeck_child(anchor.to_def_id()) {
```

Comment:
Specifically, we can't enforce this on typeck children because it would mean that the given code fails:

```rust
trait Foo {
  type Assoc;
  fn foo() -> Self::Assoc;
}

impl Foo for () {
  type Assoc = impl Sized;
  fn foo() -> Self::Assoc {
    let x = || {};
    //      ^^^^^ <- this body doesn't define `Self::Assoc`, even though it could.
    ""
  }
}

Reply:
Do you have thoughts either, @aliemjay? 🤔

Reply:
I don't remember if this landed, but I once had a commit where opaques from children got copied into their parents during borrowck. Then we could stop looking at children entirely. Well, modulo opaques that are only mentioned in the signature of children, but... oh we don't have ATPIT tests for that, do we

Reply:
We don't collect opaques for typeck children anyways:

```rust
trait Foo {
  type Assoc;
  fn foo();
}

impl Foo for () {
  type Assoc = impl Sized;
  fn foo() {
    let x = || -> Self::Assoc {
        ""
    };
  }
}
```

```
error[E0308]: mismatched types
  --> src/lib.rs:12:9
   |
9  |   type Assoc = impl Sized;
   |                ---------- the expected opaque type
...
12 |         ""
   |         ^^ expected opaque type, found `&str`
   |
   = note: expected opaque type `<() as Foo>::Assoc`
                found reference `&'static str`
```

Reply:
And I don't think we want to -- since if we don't, then we only need to check that the parent defines the given opaques.